cumulus/client/cirrus-executor/src/aux_schema.rs

```diff
@@ -22,29 +28,140 @@ fn load_decode<Backend: AuxStore, T: Decode>(
 	}
 }
 
-/// Write the execution receipt of a block to aux storage.
+/// Write the execution receipt of a block to aux storage, optionally prune the receipts that are
+/// too old.
 pub(super) fn write_execution_receipt<Backend: AuxStore, Block: BlockT>(
 	backend: &Backend,
-	block_hash: Block::Hash,
+	block_number: <<Block as BlockT>::Header as HeaderT>::Number,
 	execution_receipt: &ExecutionReceipt<Block::Hash>,
 ) -> Result<(), sp_blockchain::Error> {
-	let key = (EXECUTION_RECEIPT_KEY, block_hash).encode();
-	backend.insert_aux(&[(key.as_slice(), execution_receipt.encode().as_slice())], [])
+	let first_saved_receipt = load_decode::<_, <<Block as BlockT>::Header as HeaderT>::Number>(
+		backend,
+		EXECUTION_RECEIPT_START,
+	)?
+	.unwrap_or(block_number);
+
+	let mut new_first_saved_receipt = first_saved_receipt;
+
+	if block_number - first_saved_receipt >= PRUNING_BOUND.saturated_into() {
+		new_first_saved_receipt = block_number.saturating_sub((PRUNING_BOUND - 1).saturated_into());
+
+		let mut keys_to_delete = vec![];
+		let mut to_delete_start = first_saved_receipt;
+		while to_delete_start < new_first_saved_receipt {
+			keys_to_delete.push((EXECUTION_RECEIPT_KEY, to_delete_start).encode());
+			to_delete_start = to_delete_start.saturating_add(One::one());
+		}
+
+		backend.insert_aux(
+			&[
+				(
+					(EXECUTION_RECEIPT_KEY, block_number).encode().as_slice(),
```

Comment:
I thought I already commented on this somewhere, but ERs should be stored using block hash, not block number. Otherwise we'll get issues when handling forks of the primary chain.

Reply:
I have mentioned more in the commit message, the only way to measure time is the block number in blockchain, you can't use a hash to decide if it's too old.

Reply:
Definitely, but that doesn't mean we should use block number anyway and have issues with forks. This is exactly why the issue with leaked storage wasn't fixes quickly in BABE, it is tricky. BTW there is a draft PR (that I didn't read in much detail) that aims to deal with this: https://github.com/paritytech/substrate/pull/10748

Reply:
What's more, I'm currently not sure what's the way to handle forks in cirrus, do we just wait for a confirmation time like bitcoin or it's something else?

Reply:
No, we don't need to handle forks in any way other than following primary chain. If primary chain forks - secondary chain forks as well. If outdated branches are pruned - corresponding branches are pruned as well. This is one of the reasons why I said that this is unlike the way parachains work, execution deterministically follows primary chain.

Reply:
In order to store the receipt by hash, we need another storage to track the associated block number, otherwise, we can't decide if the receipt is too old and should be pruned. In addition, we should not delete the local receipt even it's validated correctly, otherwise if somehow the executor received the same valid receipt again, the handler will hang forever. That means the deletion of execution receipts will rely on the pruning depth only for now, but at least it won't increase without a limit.